### ISSUE TITLE
Possible OpenSSL-1.1 issue with read_ahead code

### ISSUE THREAD

**Issue by blblack** [openssl/openssl#1799]
I've run into this in the real world, using an nginx-1.11.x server built against OpenSSL-1.1.0b, in a complex production environment with live HTTPS traffic at Wikimedia.  The fallout is that our server ends up RSTing the connection under certain traffic-pattern conditions (easier to trigger with heavy client->server initial traffic), and the nginx error log saying:

```
SSL_read() failed (SSL: error:1408F119:SSL routines:ssl3_get_record:decryption failed or bad record mac) while processing HTTP/2 connection, client: N.N.N.N, server: 0.0.0.0:443
```

It's not terribly difficult for us to reproduce in the production setting, and we've had the issue reported from a wide variety of client browsers that don't share codebases.  On our production servers, I was able to narrow things down far enough that I can stop the bug by recompiling nginx with its normal call to `SSL_CTX_set_read_ahead(ssl->ctx, 1)` commented out during session setup.

From there, I tried to make a purer, offline reproduction just using the OpenSSL testsuite on the master branch.  The simplest diff I've found I can make there to reproduce something that looks related is this patch to `test/sslapitest.c` to turn on read_ahead (on the client side, in this case):

```
diff --git a/test/sslapitest.c b/test/sslapitest.c
index 4d22d8e..5303ed6 100644
--- a/test/sslapitest.c
+++ b/test/sslapitest.c
@@ -59,6 +59,8 @@ static int execute_test_large_message(const SSL_METHOD *smeth,
         goto end;
     }

+    SSL_CTX_set_read_ahead(cctx, 1);
+
     /*
      * We assume the supplied certificate is big enough so that if we add
      * NUM_EXTRA_CERTS it will make the overall message large enough. The
```

Which causes that test to fail as follows (which isn't how things are failing in production, but probably related to the same underlying issues):

```
bblack@alaxel:~/repos/gh-openssl$ make test TESTS='test_sslapi' V=1
make depend && make _tests
make[1]: Entering directory '/home/bblack/repos/gh-openssl'
make[1]: Leaving directory '/home/bblack/repos/gh-openssl'
make[1]: Entering directory '/home/bblack/repos/gh-openssl'
( cd test; \
  SRCTOP=../. \
  BLDTOP=../. \
  PERL="/usr/bin/perl" \
  EXE_EXT= \
  OPENSSL_ENGINES=.././engines \
    /usr/bin/perl .././test/run_tests.pl test_sslapi )
../test/recipes/90-test_sslapi.t .. 
1..1
./sslapitest: 91 test cases
SSL_connect() failed -1, 1
No progress made
Unable to create SSL connection
** test_large_message_tls failed **
--------
SSL_accept() failed -1, 1
SSL_connect() failed -1, 1
SSL_connect() failed -1, 1
SSL_accept() failed -1, 1
SSL_connect() failed -1, 1
SSL_accept() failed -1, 1
SSL_connect() failed -1, 1
SSL_accept() failed -1, 1
./sslapitest: 1 test failed (out of 91)
not ok 1 - running sslapitest
../util/shlib_wrap.sh ./sslapitest ../apps/server.pem ../apps/server.pem => 1

#   Failed test 'running sslapitest'
#   at ../test/recipes/90-test_sslapi.t line 21.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 

Test Summary Report
-------------------
../test/recipes/90-test_sslapi.t (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
```

Doing a little manual printf debugging into the first failure there, I find the error is the internal error marked with the comment "does not happen" at https://github.com/openssl/openssl/blob/master/ssl/record/rec_layer_s3.c#L278 .  When it reaches this point in this case, `n` is 16384, and `(rb->len - rb->offset)` (aka `rb->left`) is only 245.  It seems like the reason this "does not happen" should have been because of the move immediately above, but that move doesn't happen because `clearold` is zero in this case (otherwise it would have, as the other condition `pkt != s->rlayer.packet` is true).

My suspicion is that this issue originated with the addition of the read pipelining feature in https://github.com/openssl/openssl/commit/94777c9c86a2b2ea2726c49d6c8f61078558beba

**Comment by mattcaswell** [comment]
See #1823

**Comment by blblack** [comment]
@mattcaswell Thanks! I've rebuilt our prod packages (with nginx back to turning on read_ahead like normal, linked against OpenSSL-1.1.0b + your fix) and done some limited testing on a single server node with a real client.  I'm unable to reproduce the original problem with the RST from bad record mac errors, which I was able to do fairly easily before. :+1:

**Comment by mattcaswell** [comment]
Fixed in 0f6c9d73c.
